src/bootstrap/src/core/config/config.rs

```diff
@@ -1809,14 +1809,7 @@ impl Config {
                     }
                     target.llvm_config = Some(config.src.join(s));
                 }
-                if let Some(patches) = cfg.llvm_has_rust_patches {
-                    assert_eq!(
-                        config.submodules,
-                        Some(false),
-                        "cannot set `llvm-has-rust-patches` for a managed submodule (set `build.submodules = false` if you want to apply patches)"
-                    );
-                    target.llvm_has_rust_patches = Some(patches);
-                }
+                target.llvm_has_rust_patches = cfg.llvm_has_rust_patches;
```

Comment:
Removing the assertion isn't a suitable solution. Instead, the fix should aim to align it with what the document states.

Something like this should do the work:

```suggestion
                if let Some(patches) = cfg.llvm_has_rust_patches {
                    assert!(
                        config.submodules == Some(false) && cfg.llvm_config.is_none(),
                        "use of `llvm-has-rust-patches` is restricted to cases where either submodules are disabled or llvm-config been provided"
                    );
                    target.llvm_has_rust_patches = Some(patches);
                }
```

Reply:
Done (I'll update the PR description in a moment), though I don't really see the point of this check. Assuming that the check triggers (i.e., bootstrap is building LLVM):

* Explicitly setting `llvm-has-rust-patches = true` just makes the default behavior explicit, which doesn't really hurt.
* Explicitly setting `llvm-has-rust-patches = false` doesn't hurt either, it just disables some functionality (but as far as I am aware, things should still work just fine).

Reply:
`llvm-has-rust-patches` changes behaviors related to passing environments and flags during bootstrap. Even without the assertion, using `llvm-has-rust-patches` with submodules set to true doesn't make sense if it's not an external LLVM.